libyul/optimiser/UnusedStoreEliminator.cpp

```diff
@@ -0,0 +1,404 @@
+{
+	switch (_effect)
+	{
+	case evmasm::SemanticInformation::Effect::Read: return UnusedStoreEliminator::Effect::Read;
+	case evmasm::SemanticInformation::Effect::Write: return UnusedStoreEliminator::Effect::Write;
+	case evmasm::SemanticInformation::Effect::None: break;
+	}
+	yulAssert(false); return UnusedStoreEliminator::Effect::Read;
+}
+}
+
+vector<UnusedStoreEliminator::Operation> UnusedStoreEliminator::operationsFromFunctionCall(
+	FunctionCall const& _functionCall
+) const
+{
+	using evmasm::Instruction;
+
+	YulString functionName = _functionCall.functionName.name;
+	SideEffects sideEffects;
+	if (BuiltinFunction const* f = m_dialect.builtin(functionName))
+		sideEffects = f->sideEffects;
+	else
+		sideEffects = m_functionSideEffects.at(functionName);
+
+	optional<Instruction> instruction = toEVMInstruction(m_dialect, functionName);
+	if (!instruction)
+	{
+		vector<Operation> result;
+		// Unknown read is worse than unknown write.
+		if (sideEffects.memory != SideEffects::Effect::None)
+			result.emplace_back(Operation{Location::Memory, Effect::Read, {}, {}});
+		if (sideEffects.storage != SideEffects::Effect::None)
+			result.emplace_back(Operation{Location::Storage, Effect::Read, {}, {}});
+		return result;
+	}
+
+	using evmasm::SemanticInformation;
+
+	return util::applyMap(SemanticInformation::readWriteOperations(*instruction),
+		[&](SemanticInformation::Operation const& _op) -> Operation
+	{
+		yulAssert(!(_op.lengthParameter && _op.lengthConstant));
+		Operation ourOp{translateLocation(_op.location), translateEffect(_op.effect), {}, {}};
+		if (_op.startParameter)
+			ourOp.start = identifierNameIfSSA(_functionCall.arguments.at(*_op.startParameter));
+		if (_op.lengthParameter)
+			ourOp.length = identifierNameIfSSA(_functionCall.arguments.at(*_op.lengthParameter));
+		if (_op.lengthConstant)
+			switch (*_op.lengthConstant)
+			{
+			case 1: ourOp.length = YulString(one); break;
+			case 32: ourOp.length = YulString(thirtyTwo); break;
+			default: yulAssert(false);
+			}
+		return ourOp;
+	});
+}
+
+void UnusedStoreEliminator::applyOperation(UnusedStoreEliminator::Operation const& _operation)
+{
+	for (auto& [statement, state]: m_stores[YulString{}])
+		if (state == State::Undecided)
+		{
+			Operation const& storeOperation = m_storeOperations.at(statement);
+			if (_operation.effect == Effect::Read && !knownUnrelated(storeOperation, _operation))
+				state = State::Used;
+			else if (_operation.effect == Effect::Write && knownCovered(storeOperation, _operation))
+				state = State::Unused;
+		}
+}
+
+bool UnusedStoreEliminator::knownUnrelated(
+	UnusedStoreEliminator::Operation const& _op1,
+	UnusedStoreEliminator::Operation const& _op2
+) const
+{
+	KnowledgeBase knowledge(m_dialect, m_ssaValues);
+
+	if (_op1.location != _op2.location)
+		return true;
+	if (_op1.location == Location::Storage)
+	{
+		if (_op1.start && _op2.start)
+			return knowledge.knownToBeDifferent(*_op1.start, *_op2.start);
```

Comment:
Since we did set the lengths to constant 1 for these, maybe we should assert that they're there and that they are one here.